src/sinks/kafka/sink.rs

```diff
@@ -0,0 +1,140 @@
+use async_trait::async_trait;
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use rdkafka::consumer::{BaseConsumer, Consumer};
+use rdkafka::error::KafkaError;
+use rdkafka::producer::FutureProducer;
+use rdkafka::ClientConfig;
+use snafu::{ResultExt, Snafu};
+use std::convert::TryFrom;
+use std::num::NonZeroUsize;
+use tokio::time::Duration;
+use vector_core::buffers::Acker;
+
+#[derive(Debug, Snafu)]
+pub enum BuildError {
+    #[snafu(display("creating kafka producer failed: {}", source))]
+    KafkaCreateFailed { source: KafkaError },
+    #[snafu(display("invalid topic template: {}", source))]
+    TopicTemplate { source: TemplateParseError },
+}
+
+pub struct KafkaSink {
+    encoding: EncodingConfig<Encoding>,
+    acker: Acker,
+    service: KafkaService,
+    topic: Template,
+    key_field: Option<String>,
+    headers_field: Option<String>,
+}
+
+pub fn create_producer(
+    client_config: ClientConfig,
+) -> crate::Result<FutureProducer<KafkaStatisticsContext>> {
+    let producer = client_config
+        .create_with_context(KafkaStatisticsContext)
+        .context(KafkaCreateFailed)?;
+    Ok(producer)
+}
+
+impl KafkaSink {
+    pub(crate) fn new(config: KafkaSinkConfig, acker: Acker) -> crate::Result<Self> {
+        let producer_config = config.to_rdkafka(KafkaRole::Producer)?;
+        let producer = create_producer(producer_config)?;
+
+        Ok(KafkaSink {
+            headers_field: config.headers_field,
+            encoding: config.encoding,
+            acker,
+            service: KafkaService::new(producer),
+            topic: Template::try_from(config.topic).context(TopicTemplate)?,
+            key_field: config.key_field,
+        })
+    }
+
+    async fn run_inner(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let request_builder_concurrency_limit = NonZeroUsize::new(50);
+
+        let request_builder = KafkaRequestBuilder {
+            topic: self.topic,
+            key_field: self.key_field,
+            headers_field: self.headers_field,
+        };
+
+        let sink = input
+            .request_builder(
+                request_builder_concurrency_limit,
+                request_builder,
+                self.encoding,
+                Compression::None,
+            )
+            .filter_map(|request| async move {
+                match request {
+                    Ok(req) => Some(req),
+                    Err(RequestBuilderError::EncodingError(e)) => {
+                        error!("Failed to encode Kafka request: {:?}.", e);
+                        None
+                    }
+                    Err(RequestBuilderError::SplitError(error)) => {
+                        emit!(&TemplateRenderingFailed {
+                            error,
+                            field: Some("topic"),
+                            drop_event: true,
+                        });
+                        None
+                    }
+                }
+            })
```

Comment:
I'm a little iffy on changing the `RequestBuilder` trait, in no small part because I'm currently trying to think about how to make it more generic overall, but primarily because we're adding a path that is fairly niche: the lion's share of sinks will be batching themselves which is where we can already inject fallibility i.e. when template rendering fails.

Additionally, given the fact you're not compressing or encoding more than one event at a time makes me think that we'd be better off putting the request building logic into a single function that returns `Option<KafkaRequest>` and then gets run via `filter_map`.

We'd be executing a lot of machinery, namely spawning Tokio tasks and going through all of the request builder hierarchy, on every single event.  If we push it through `filter_map`, we get a lot more locality and the sink task should be able to run for longer.

Reply:
This is also something I feel iffy about for the reasons Toby laid out.